### PR TITLE
Fix/Morph client

### PR DESCRIPTION
**Pull request by carpawell**

### CODE REVIEW

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3682

**Comment by sami-nspcc** [comment]
I am running integration tests

File: pkg/morph/client/constructor.go [hunk 1]
```diff
@@ -158,6 +158,12 @@ func (c *Client) newCli(endpoint string) (*rpcclient.WSClient, *actor.Actor, *ne
 		return nil, nil, nil, fmt.Errorf("WS client creation: %w", err)
 	}
 
+	defer func() {
+		if err != nil {
+			cli.Close()
+		}
```
Comment (fyrchik) [inline pkg/morph/client/constructor.go:164]: Why don't we close it in any case (before or after `Init`)?

File: pkg/morph/client/notary.go [hunk 2]
```diff
@@ -176,7 +175,7 @@ func (c *Client) DepositNotary(amount fixedn.Fixed8, delta uint32) (res util.Uin
 		big.NewInt(int64(amount)),
 		[]interface{}{c.acc.PrivateKey().GetScriptHash(), till})
 	if err != nil {
-		if !errors.Is(err, neorpc.ErrAlreadyExists) {
+		if !alreadyExistsError(err) {
```
Comment (fyrchik) [inline pkg/morph/client/notary.go:178]: Why didn't it work the old way?

File: pkg/morph/client/constructor.go (hunk 1 above)
Comment (carpawell) [reply pkg/morph/client/constructor.go:164]: cause i did not want to write `cli.Close()` in every `err != nil`, the only reason. we should close the cli even if `actor`
`gas` creation has failed because we will try to open a new ws conn anyway

i can change it to the repeated calls in every `if` branch if you want

File: pkg/morph/client/notary.go (hunk 2 above)
Comment (carpawell) [reply pkg/morph/client/notary.go:178]: that is a server error, i did not find any code that returns _exactly_ that err on the client side, did i miss anything?

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3686

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/DXEX2FrLGvYsLc4fvvLURzsFCnuR533wEafdMe6DVTnQ">link</a>. Untar and use `allure open` to watch the report

File: pkg/morph/client/notary.go (hunk 2 above)
Comment (fyrchik) [reply pkg/morph/client/notary.go:178]: I think it doesn't matter here https://github.com/nspcc-dev/neo-go/blob/master/pkg/neorpc/errors.go#L131

File: cmd/neofs-node/morph.go [hunk 3]
```diff
@@ -133,7 +133,7 @@ func makeNotaryDeposit(c *cfg) (util.Uint256, error) {
 
 		// gasDivisor defines what part of GAS balance (1/gasDivisor)
 		// should be transferred to the notary service
-		gasDivisor = 2
+		gasDivisor = 3
```
Comment (fyrchik) [inline cmd/neofs-node/morph.go:136]: What about all other, more frequent cases?

**Review by fyrchik — CHANGES_REQUESTED** [review]

File: pkg/morph/client/notary.go (hunk 2 above)
Comment (carpawell) [reply pkg/morph/client/notary.go:178]: in fact, it seems like an unreachable code, cause all the TXs have their own `nonce`, i checked that

so ok, just dropped that commit

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/Gd2qYUEPN3Zv1Q7VzaeQAJaKGFEoTdCUe9CcdPDBWJKD">link</a>. Untar and use `allure open` to watch the report

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3701

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3704

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/ACH9GV6aTZopSiiZ4sMABkoPgZ7rhoGgZKqgnWVg6wMK">link</a>. Untar and use `allure open` to watch the report

File: CHANGELOG.md [hunk 4]
```diff
@@ -38,6 +38,8 @@ Changelog for NeoFS Node
 - Panic in IR when performing HEAD requests (#2069)
 - Write-cache flush duplication (#2074)
 - Ignore error if a transaction already exists in a morph client (#2075)
+- Pack arguments of `setPrice` invocation during contract update (#2078)
```
Comment (fyrchik) [inline CHANGELOG.md:43]: wut?

File: CHANGELOG.md [hunk 5]
```diff
@@ -41,6 +41,7 @@ Changelog for NeoFS Node
 - Pack arguments of `setPrice` invocation during contract update (#2078)
 - Closing `neo-go` WS clients on shutdown and switch processes (#2080)
 - Making notary deposits with a zero GAS balance (#2080)
+- Notary requests on the shutdown (#2075)
```
Comment (fyrchik) [inline CHANGELOG.md:46]: `on shutdown`

File: cmd/neofs-node/morph.go (hunk 3 above)
Comment (fyrchik) [reply cmd/neofs-node/morph.go:136]: I don't think we need this hotfix in master -- it "fixes" a rather peculiar and artificial scenario.
Let's discuss and fix it properly in another task.

Comment (carpawell) [reply cmd/neofs-node/morph.go:136]: Hm, `2` is still a magic value anyway cause nobody knows what exact value is perfect for a deposit but it certainly does not allow making two deposits (caused for any reason) in a row

dropped

File: CHANGELOG.md (hunk 4 above)
Comment (carpawell) [reply CHANGELOG.md:43]: rebase mistake, dropped

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3732

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3733

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3734

**Comment by sami-nspcc** [comment]
I am running integration tests

**Review by acid-ant — APPROVED** [review]

File: pkg/services/object/get/get_test.go [hunk 6]
```diff
@@ -963,7 +963,7 @@ func TestGetRemoteSmall(t *testing.T) {
 				addr.SetObject(oidtest.ID())
 
 				srcObj := generateObject(addr, nil, nil)
-				srcObj.SetPayloadSize(10)
+				srcObj.SetPayloadSize(11)
```
Comment (fyrchik) [inline pkg/services/object/get/get_test.go:966]: What is the reason for this change?

File: pkg/services/object/get/assemble.go [hunk 7]
```diff
@@ -203,7 +203,9 @@ func (exec *execCtx) buildChainInReverse(prev oid.ID) ([]oid.ID, []objectSDK.Ran
 
 	// fill the chain end-to-start
 	for withPrev {
-		if exec.curOff < from {
+		// check that only for "range" requests,
+		// for `GET` it stops via the false `withPrev`
+		if seekRng != nil && exec.curOff <= from {
```
Comment (fyrchik) [inline pkg/services/object/get/assemble.go:208]: Did we have a bug here?

**Review by fyrchik — CHANGES_REQUESTED** [review]

File: pkg/services/object/get/get_test.go (hunk 6 above)
Comment (carpawell) [reply pkg/services/object/get/get_test.go:966]: incorrect test: parent length is `10`, the first child obj's len is `10`, test expects another request to the child obj (it will not happen after the fix)

File: pkg/services/object/get/assemble.go (hunk 7 above)
Comment (carpawell) [reply pkg/services/object/get/assemble.go:208]: yes, if a child obj's boundary is equal to the original `Range`'s offset, a node does another useless `[X, 0]` `Range` request

fixed by `<` => `<=` change

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3741

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/4pF3GGJKuTt7zz1ideDxNTKyq8zJJvBmrgbibVu4r5ts">link</a>. Untar and use `allure open` to watch the report

**Review by fyrchik — APPROVED** [review]

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/6NmDB3mu4jZQdHbLQU7RW6HN1Uaqm6rdZNbsinFvYKpB">link</a>. Untar and use `allure open` to watch the report
